src/main/java/ch/njol/skript/effects/EffPotion.java

```diff
@@ -35,144 +29,70 @@
-			potionEffects = (Expression<PotionEffect>) exprs[0];
-			entities = (Expression<LivingEntity>) exprs[1];
-		} else if (apply) {
-			potions = (Expression<PotionEffectType>) exprs[0];
-			tier = (Expression<Number>) exprs[1];
-			entities = (Expression<LivingEntity>) exprs[2];
-			duration = (Expression<Timespan>) exprs[3];
-		} else {
-			potions = (Expression<PotionEffectType>) exprs[0];
-			entities = (Expression<LivingEntity>) exprs[1];
-		}
-		
-		// Ambience and particles
-		switch (matchedPattern) {
-			case 0:
-				ambient = false;
-				particles = true;
-				break;
-			case 1:
-				ambient = true;
-				particles = true;
-				break;
-			case 2:
-				ambient = false;
-				particles = false;
-				break;
-		}
-		
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		boolean first = matchedPattern == 0;
+		potionEffects = (Expression<PotionEffect>) exprs[first ? 0 : 1];
+		entities = (Expression<LivingEntity>) exprs[first ? 1 : 0];
+		duration = (Expression<Timespan>) exprs[2];
 		return true;
 	}
 	
 	@Override
-	protected void execute(final Event e) {
-		if (potionEffect) {
-			for (LivingEntity livingEntity : entities.getArray(e)) {
-				PotionEffect[] potionEffects = this.potionEffects.getArray(e);
-				PotionEffectUtils.addEffects(livingEntity, potionEffects);
-			}
-		} else {
-			final PotionEffectType[] ts = potions.getArray(e);
-			if (ts.length == 0)
-				return;
-			if (!apply) {
-				for (LivingEntity en : entities.getArray(e)) {
-					for (final PotionEffectType t : ts)
-						en.removePotionEffect(t);
-				}
-				return;
-			}
-			int a = 0;
-			if (tier != null) {
-				final Number amp = tier.getSingle(e);
-				if (amp == null)
-					return;
-				a = amp.intValue() - 1;
-			}
-			int d = DEFAULT_DURATION;
-			if (duration != null) {
-				final Timespan dur = duration.getSingle(e);
-				if (dur == null)
-					return;
-				d = (int) (dur.getTicks_i() >= Integer.MAX_VALUE ? Integer.MAX_VALUE : dur.getTicks_i());
-			}
-			for (final LivingEntity en : entities.getArray(e)) {
-				for (final PotionEffectType t : ts) {
-					int duration = d;
-					if (!replaceExisting) {
-						if (en.hasPotionEffect(t)) {
-							for (final PotionEffect eff : en.getActivePotionEffects()) {
-								if (eff.getType() == t) {
-									duration += eff.getDuration();
-									break;
-								}
-							}
-						}
-					}
-					en.addPotionEffect(new PotionEffect(t, duration, a, ambient, particles), true);
-				}
+	protected void execute(Event e) {
+		PotionEffect[] potionEffects = this.potionEffects.getArray(e);
+
+		// Change duration for some backwards compatibility with older Skript versions
+		if (duration != null) {
+			Timespan timespan = duration.getSingle(e);
+			if (timespan != null) {
+				int ticks = (int) timespan.getTicks_i();
+				for (int i = 0; i < potionEffects.length; i++)
+					potionEffects[i] = potionEffects[i].withDuration(ticks);
```

Comment:
this can throw an ArrayIndexOutOfBoundsException if the array is empty (aka of length 0, and getArray will pass empty arrays for null elements), so you gotta check if potionEffects.length != 0 before this

Reply:
I don't think it would since 0 < 0 wouldn't pass

I am reworking this anyways though like we discussed on Discord